From: "Jeroen van Meeuwen (Kolab Systems)" <vanmeeuwen@kolabsys.com>
Date: Tue, 3 Aug 2010 10:19:11 +0200
Subject: Normalize the authentication ID

By normalize, it is intended that;

    1) Authentication IDs all can be lowercased for more accurate
       comparison without being volatile to, say, user error, and
    2) Any leading or trailing blank space can be stripped
---
 imap/global.c    |  2 ++
 lib/auth_unix.c  | 18 ++++++++++++++++++
 lib/imapoptions  |  5 +++++
 lib/libcyr_cfg.c |  4 ++++
 lib/libcyr_cfg.h |  2 ++
 5 files changed, 31 insertions(+)

diff --git a/imap/global.c b/imap/global.c
index b038d3d..cd51b78 100644
--- a/imap/global.c
+++ b/imap/global.c
@@ -277,6 +277,8 @@ EXPORTED int cyrus_init(const char *alt_config, const char *ident, unsigned flag
 				  config_getswitch(IMAPOPT_UNIX_GROUP_ENABLE));
 	libcyrus_config_setswitch(CYRUSOPT_USERNAME_TOLOWER,
 				  config_getswitch(IMAPOPT_USERNAME_TOLOWER));
+	libcyrus_config_setswitch(CYRUSOPT_NORMALIZEUID,
+				  config_getswitch(IMAPOPT_NORMALIZEUID));
 	libcyrus_config_setswitch(CYRUSOPT_SKIPLIST_UNSAFE,
 				  config_getswitch(IMAPOPT_SKIPLIST_UNSAFE));
 	libcyrus_config_setstring(CYRUSOPT_TEMP_PATH,
diff --git a/lib/auth_unix.c b/lib/auth_unix.c
index 1f034da..914ea0b 100644
--- a/lib/auth_unix.c
+++ b/lib/auth_unix.c
@@ -150,9 +150,11 @@ static char allowedchars[256] = {
 static const char *mycanonifyid(const char *identifier, size_t len)
 {
     static char retbuf[81];
+    char backup[81];
     struct group *grp;
     char *p;
     int username_tolower = 0;
+    int ic,rbc;
 
     if (!len) len = strlen(identifier);
     if (len >= sizeof(retbuf)) return NULL;
@@ -194,6 +196,22 @@ static const char *mycanonifyid(const char *identifier, size_t len)
 	}
     }
 
+    if( (libcyrus_config_getswitch(CYRUSOPT_NORMALIZEUID) == 1) ) {
+        strcpy(backup,retbuf);
+       /* remove leading blanks */
+       for(ic=0; isblank(backup[ic]); ic++);
+       for(rbc=0; backup[ic]; ic++) {
+            retbuf[rbc] = ( isalpha(backup[ic]) ?
+                 tolower(backup[ic]) : backup[ic] );
+            rbc++;
+       }
+       retbuf[rbc] = '\0';
+       /* remove trailing blanks */
+       for(--rbc; isblank(retbuf[rbc]); rbc--) {
+            retbuf[rbc] = '\0';
+       }
+    }
+
     return retbuf;
 }
 
diff --git a/lib/imapoptions b/lib/imapoptions
index d80f081..5f7c523 100644
--- a/lib/imapoptions
+++ b/lib/imapoptions
@@ -2008,6 +2008,11 @@ product version in the capabilities */
 /* The absolute path to the zoneinfo db file.  If not specified,
    will be confdir/zoneinfo.db */
 
+{ "normalizeuid", 0, SWITCH }
+/* Lowercase uid and strip leading and trailing blanks. It is recommended
+   to set this to yes, especially if OpenLDAP is used as authentication
+   source. */
+
 /*
 .SH SEE ALSO
 .PP
diff --git a/lib/libcyr_cfg.c b/lib/libcyr_cfg.c
index 57b7b24..58ac088 100644
--- a/lib/libcyr_cfg.c
+++ b/lib/libcyr_cfg.c
@@ -156,6 +156,10 @@ static struct cyrusopt_s cyrus_options[] = {
       CFGVAL(long, 1),
       CYRUS_OPT_SWITCH },
 
+    { CYRUSOPT_NORMALIZEUID,
+      CFGVAL(long, 1),
+      CYRUS_OPT_SWITCH },
+
     { CYRUSOPT_LAST, { NULL }, CYRUS_OPT_NOTOPT }
 };
 
diff --git a/lib/libcyr_cfg.h b/lib/libcyr_cfg.h
index 3983ccd..8f48422 100644
--- a/lib/libcyr_cfg.h
+++ b/lib/libcyr_cfg.h
@@ -111,6 +111,8 @@ enum cyrus_opt {
     CYRUSOPT_SQL_USESSL,
     /* Checkpoint after every recovery (OFF) */
     CYRUSOPT_SKIPLIST_ALWAYS_CHECKPOINT,
+    /* Lowercase uid and strip leading and trailing blanks (OFF) */
+    CYRUSOPT_NORMALIZEUID,
 
     CYRUSOPT_LAST
     
